tsl/src/gapfill/planner.c

```diff
@@ -144,67 +174,121 @@ gapfill_plan_create(PlannerInfo *root, RelOptInfo *rel, struct CustomPath *path,
-	 * above aggregation node.
-	 */
-	if (root->upper_targets[UPPERREL_WINDOW] &&
-		list_length(root->upper_targets[UPPERREL_WINDOW]->exprs) !=
-			list_length(root->upper_targets[UPPERREL_GROUP_AGG]->exprs))
-		ereport(ERROR,
-				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("gapfill functionality with window functions not supported")));
+	return &cscan->scan.plan;
+}
+
+static CustomPathMethods gapfill_path_methods = {
+	.CustomName = "GapFill",
+	.PlanCustomPath = gapfill_plan_create,
+};
 
-	for (i = 0; i < list_length(tlist); i++)
+static bool
+gapfill_expression_walker(Expr *node, bool (*walker)(), gapfill_walker_context *context)
+{
+	context->count = 0;
+	context->call.node = NULL;
+
+	return (*walker)((Node *) node, context);
+	return expression_tree_walker((Node *) list_make1(node), walker, context);
+}
+
+/*
+ * Build expression lists for the gapfill node and the node below.
+ * All marker functions will be top-level function calls in the
+ * resulting gapfill node expression list and will not be included in
+ * the subpath expression list
+ */
+static void
+gapfill_build_pathtarget(PathTarget *pt_upper, PathTarget *pt_path, PathTarget *pt_subpath)
+{
+	ListCell *lc;
+	int i = -1;
+
+	foreach (lc, pt_upper->exprs)
 	{
-		tle = list_nth(tlist, i);
-		if (root->upper_targets[UPPERREL_WINDOW])
+		Expr *expr = lfirst(lc);
+		gapfill_walker_context context;
+		i++;
+
+		/* check for locf/interpolate calls */
+		gapfill_expression_walker(expr, marker_function_walker, &context);
+		if (context.count > 1)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("multiple interpolate/locf function calls per column not supported")));
+
+		if (context.count == 1)
 		{
-			expr = list_nth(root->upper_targets[UPPERREL_WINDOW]->exprs, i);
-			if (IsA(expr, WindowFunc))
-				expr = linitial(castNode(WindowFunc, expr)->args);
-			tl_exprs = lappend(tl_exprs, expr);
+			/*
+			 * marker needs to be toplevel for now unless we have a projection capable
+			 * node above gapfill node
+			 */
+			if (expr != context.call.expr && !contain_window_function((Node *) expr))
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("%s must be toplevel function call",
+								get_func_name(context.call.func->funcid))));
+
+			/* if there is an aggregation it needs to be a child of the marker function */
+			if (contain_agg_clause((Node *) expr) &&
+				!contain_agg_clause(linitial(context.call.func->args)))
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("aggregate functions must be below %s",
+								get_func_name(context.call.func->funcid))));
+
+			if (contain_window_function(context.call.node))
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("window functions must not be below %s",
+								get_func_name(context.call.func->funcid))));
+
+			add_column_to_pathtarget(pt_path, context.call.expr, pt_upper->sortgrouprefs[i]);
```

Comment:
add comment that adding pathtarget including marker to gapfill node and the plain aggregate to the nodes below.